### DISCUSSION TITLE
Custom Menu Items [for LLM Ollama / RAG, via Open-Web UI on port 3030]

### DISCUSSION THREAD

**Discussion by CrimsonPaladinActual** [iiab/iiab#3799] [Q&A] [answered]
I've dug through all the documentation, and for life of me I can't make this work. I'm trying to create a new menu icon and definition for a custom service running on the device utilizing port 3030. I've tried editing clones of several standard menu items, but they all start with residual characters after localhost, for example: 10.10.10.10/:3030. Is there a way to do this from the admin portal or am I going to have to go into the file system and manually edit files?

**Comment by holta** [comment]
> Is there a way to do this from the admin portal

I don't think so.

@tim-moody might possibly have ideas in coming days?

PS what are you running on Port 3030?

**Comment by CrimsonPaladinActual** [comment]
I'm running Ollama and using Open-Web UI at port 3030. It's on a Raspberry Pi 4, so nothing crazy, but I'm playing around with its RAG implementation with regards to IIAB's reference material. Like using it to be able to query a PDF version of the IIAB FAQ. I already made a custom menu item for Gitea; just trying to pretty things up a bit for the non-technical folks.

**Reply by holta** [reply]
Promising !

Are you in touch with @Abdulrahman392011 doing similar?

- #3796

**Reply by CrimsonPaladinActual** [reply]
I am not; I'm planning on looking at his work this evening.

**Reply by holta** [reply]
Great news 💯 

If you email holt@unleashkids.org, I can put you in touch with him directly, so you can learn from each other if that helps 🏹

**Reply by holta** [reply]
> I'm running Ollama and using Open-Web UI at port 3030.

Interesting!  I wasn't aware of Open-Web UI (https://OpenWebUI.com, https://github.com/open-webui) == "friendly self-hosted WebUI designed to operate entirely offline..."

- Formerly called "Ollama WebUI"
- Supports various LLM runners, including Ollama and OpenAI-compatible APIs
- ["Extensible, feature-rich"](https://github.com/open-webui/open-webui#open-webui-formerly-ollama-webui-)

**Reply by CrimsonPaladinActual** [reply]
It really is a cool way to demonstrate LLMs alongside the awesome IIAB platform. To make it work with Raspberry Pi 4 hardware I had to use a prior Open-Web UI image, as the current image seems to have changed some things in regard to Raspberry Pi 5 CPU architecture. As an example, this is my Docker compose.yaml:

services:
  open-webui:
    image: ghcr.io/open-webui/open-webui:git-c74e7df
    container_name: open-webui
    volumes:
      - ./data:/app/backend/data
    ports:
      - 3030:8080
    extra_hosts:
      - host.docker.internal:host-gateway
    restart: unless-stopped

This works on a Raspberry Pi 4 running IIAB.

**Reply by Abdulrahman392011** [reply]
One day in the near future we will look back at the 2 billion parameters models and laugh but for now we have to stay within the boundaries of our cpu and the raspberry pi runs ollama on cpu. If I were you I'll look into gpu parallelism and accelerated computing. Once again i remind you that pi 5 is slower than my 9 year's old laptop which isn't convenient in running models to say the least, not to mention that you're talking about pi 4 not 5. In the recent years running models on cpu will be obsolete but for now it is what it is.

**Reply by Abdulrahman392011** [reply]
To put some prospective, about 70% of the die area in the m chips of apple is dedicated to the gpu. This is how important the gpu is. Running models on the cpu is almost arcadic.

**Reply by holta** [reply]
There's some claim that Raspberry Pi 4 and 5 both contain some onboard neural/AI silicon (chip?) helpful for AI.

_But possibly that's not exactly helpful in this case?_

**Reply by Abdulrahman392011** [reply]
If it does then ollama doesn't support it cause i just watched a video of a guy running some models on it and it was a little slower than my laptop. But the video was about 5 month old. Just when the pi5 was announced so ollama might have supported it by now. All guesses!

**Reply by Abdulrahman392011** [reply]
I do hope they do,  it will increase the value of the pi significantly. I mean if you can pay less than a 100 dollars and get decent performance. But i think it's too good to be true kinda scenario. At least by nowadays standards.

**Reply by Abdulrahman392011** [reply]
Alrigh, so i done a quick search and it turns out you're kinda right. The pi5 don't have a neural engine itself, however you can add one to it for 13 tops of performance which is way better than running on the cpu. That being said there's no guarantee that ollama supports it.

**Comment by CrimsonPaladinActual** [comment]
@Abdulrahman392011 I totally agree; typically I use GPU acceleration. In this particular case it is just a cool proof-of-concept demonstration. 

@holta there are AI-accelerator add-ons for the Pi, but as far as I know they are not super beneficial for running general LLMs.

**Comment by Abdulrahman392011** [comment]
https://hailo.ai/products/generative-ai-accelerators/hailo-10h-m-2-generative-ai-acceleration-module/#hailo10m2-overview

The link above have a module that fits on the pi5 and gives 40tops of performance. It works with llms. Maybe it can work with the pi4 as well. Not sure yet

**Comment by Abdulrahman392011** [comment]
Hey, so i found something that could be a deal breaker depending on how you look at it. Basically the hailo chip can't run ollama models. You have to get the original models in their floating state and then quantiz them with hailo quantizer. This means updating the models is gonna be much harder and also using model for vision and audio will be more complicated. They also have their own api so any code written for ollama will need some modification and testing before usage. That being said they still are a viable option for mass production but for prototyping I would vote against them. It over complicate things unnecessarily. It makes much more sense to take the motherboard out of a laptop that have multiple pcp design for its motherboard like the Lenovo yoga new laptop with Intel's lunar chip. That is if you're prototyping for some product you're putting together. It will be expensive but it's a one time payment. When things are working then you can build using the hailo chip and return the motherboard back to the laptop and sell it used for half of what you bought it for or something. Also did i mention how hard it is to get your hands on one of these hailo chips. They prioritize the product manufacturers over hobbyists and R&D. I don't blame them, they have contracts and deadlines. Also anything extra that they manufacturers becomes outdated very vast if not sold fast so they don't really gamble with their company.

**Comment by Abdulrahman392011** [comment]
Hi guys, so i was looking into the hailo h10 chip and it turns out that raspberry pi the company  doesn't actually make a hat that integrates with the pi with that chip on board.
About 3 months ago they made a hat with another hailo chip, 8L if i recal was it's name. And basically that chip can run vision models and convolution models only but not transformer based model like text based models and embedding models. Bottom line is, you can't use the pi yet, hopefully they'll make another hat with the h10. Hopefully they'll do it in the near future. But just guessing, no actual promises found anywhere online.

Ps. I contacted hailo and they only make the chip not the actual hat. That is raspberry pi responsibility. To design the board (hat) and buy the chip from hailo and put it on their board (hat). 

I don't want to sound judgmental but why did raspberry start with vision? It makes more sense to go text models first. They're much more common and intriguing to try out.

**Reply by CrimsonPaladinActual** [reply]
If I had to guess it would be due to the amount of integrated projects being generated by the maker community using AI for things such as object detection, range estimation, etc. As we discussed previously, LLM use on the Pi is rudimentary at best and even the larger consumer-grade GPUs struggle with producing responses that are actually useful due to applicable model size. Between RAM constraints, power utilization, and processing power, the decision makers probably decided to pursue module development that is more fruitful given current technological constraints.

**Reply by Abdulrahman392011** [reply]
So i actually asked that question to raspberry and no answer was provided in that regard they just told me that they will send my feedback to the department responsible and feedback is helpful. 
I also told them to have a survey before making decisions. 

Believe it or not the product that we have in our hands now was decided for about two years ago. Back then LLMs weren't a thing. And convolutional models were the only one commonly known

**Comment by Abdulrahman392011** [comment]
Correction Lenovo yoga aura not asus is the laptop with three pcb and lunar intel chip. The one that i mentioned earlier. You can see it opened in this video at 3:47 

https://m.youtube.com/watch?v=STpFf-cdCSM

**Comment by tim-moody** [comment, marked as answer]
back to the original issue, we have no menu item type for ports, though we have considered adding one in the past.

**Reply by CrimsonPaladinActual** [reply]
Thank you for clearing it up; glad I wasn't missing something. If it isn't too heavy of a lift, I think this would be a nice feature in future versions. Right now it isn't a huge deal, but as the Pi family grows more capable this may be an easy way for admins to integrate additional services. It would not only make them easy for users to access, but having a menu item for them helps to advertise their availability to users accessing the IIAB.

**Comment by tim-moody** [comment]
It's not a big deal, but other things are higher on my list at the moment. We originally went down the road of having a separate type of each of the several apps that use ports, so now having a type for all the rest that use ports is a little unnatural, but better than nothing I guess.

If anyone wants to create a PR, I'm open to it.

**Reply by Abdulrahman392011** [reply]
How about linking a terminal to a html page. It's insecure but it will look nicer than using a separate window. And it's universal and can be used if you're the only one using the machine.

We can link it to a certain command and terminate it when ctrl d was pressed if it's running python or have different termination hot keys depending on the file type that's being executed.

Also the user can add as many of them as he wants. Linking each to a certain initiate and terminate commands.

I don't know what i am taking about so feel free to be dismissive. Just brainstorming.

**Comment by tim-moody** [comment]
>We used Ollama SLMs with a fully embedded Wikipedia ZIM file

How do you embed the ZIM? Is it not a compressed binary?

**Reply by OscSanto** [reply]
It is, but the libzim library decompresses ZIM binary into HTML format for each article within. Then we do iteration and filtering through each article, alongside a strategy to chunk. After chunking, we embed. 

What we used to do (and see often still) is rely on Kiwix searchbar function to retrieve relevant articles based on user query. This often led to poor results due to its lexical only search signals and lack of ZIM crossover search results. We would then take top ~15 article matches and proceed to make multiple HTTP calls to scrape each article, then chunk and embed. 

To embed 15+ articles PER user query takes a noticable amount of time to complete on rpi5.

With pre-embedding & libzim we saved considerable time, no more round HTTP calls or lexical only searching, and we gain cross-ZIM semantic search.

**Comment by tim-moody** [comment]
Yes. In your code I see you scrape. There is zimdump that could be faster. https://wiki.openzim.org/wiki/Zimdump
